LTP: Passing ext4 block device to enclave

shaikshavali1 opened the pull request:
This test case causing oom killer to be invoked and causing
kernel panic. This is because the test case invokes test framework
function "tst_acquire_device" which creates a 256MB of .img file
and binds with a loop device. This is not supported because the
default size of LKL memory is set to 32M. This is kept low due to
the limited size of EPC (Enclave page cache) and to avoid page
cache being swapped out.
Also, the test case invokes "tst_mkfs" framework function which
inturn invokes mkfs utility command with the help of system() syscall.
It is recommended to use the root file system, but, this test case
needs a read-only filesystem to test one of the sub-test cases.
we don't have a read-only file system mounted in sgx-lkl.

To support the test cases which need a block device. A formatted
ext4 filesystem image is created and passed
to enclave as an extended disk. The test case passes the details about this
new block disk image in the application configuration file. The test framework
will pick this image details in the environment variable and use it in the test.
Test case changes link: lsds/ltp#49, lsds/ltp#59.

This PR also revert the PR:#741
This #741, creates a filesystem image in the master image.
because of this, each individual test case needs to be modified.

Commit:
Don't overwrite LTP tests when doing make clean

Prior to this change when doing `make clean` in the `ltp` directory,
it would resync the ltp git submodule. This is highly problematic
for any workflow where you would need to work with a different commit
of the ltp submodule or if you want to make changes in tree to test.

For example, if debugging a problem with an ltp test.

You modify the ltp test.
Run make clean and go to run the given test.

Prior to this change, all your modifications to the ltp test would be
overwritten.

As of this commit, the ltp submodule will only be sync'd if it hasn't
already been pulled down. This should keep existing workflows operating
while also making submodule checkout for ltp work the same as those for
sgx-lkl-musl etc.

Commit:
Use fresh filesystem for each LTP test invocation

This change appears to have been missed during a previous PR.

Commit:
Undo LTP patches related to #224

Issue #224 was fixed by #723.

Closes #224

Commit:
Unify LTP Makefiles

Prior to this commit, there were two sets of almost identical "Makefiles"
for LTP.

batch.mk that was used for the Makefiles in ltp-batch* and Makefile that
is currently unused in CI, but can be used to run tests manually from the
LTP directory.

This commit changes to have both sets of "Makefiles" use batch.mk by
merging the few small differences in relatives paths into batch.mk and
checking for the files are known locations and setting up accordindly.

shaikshavali1 committed:
Revert "LTP: Added a support to mount file system in tests (#741)"

This reverts commit a401b9149d74e5be2a39fdca71a7601aaabdaf41.

shaikshavali1 committed:
Revert "LTP: Added a support to mount file system in tests (#741)"

This reverts commit a401b9149d74e5be2a39fdca71a7601aaabdaf41.

shaikshavali1 committed:
LTP: Pass ext4 block device to enclave

This test case causing oom killer to be invoked and causing
kernel panic. This is because the test case invokes test framework
function "tst_acquire_device" which creates a 256MB of .img file
and binds with a loop device. This is not supported because the
default size of LKL memory is set to 32M. This is kept low due to
the limited size of EPC (Enclave page cache) and to avoid page
cache being swapped out.
Also, the test case invokes "tst_mkfs" framework function which
inturn invokes mkfs utility command with the help of system() syscall.
It is recommended to use root file system, but, this test case
needs a read-only filesystem to test one of the sub-test cases.
we don't have a read-only file system mounted in sgx-lkl.

To support the test cases which need a block device a formatted
ext4 filesystem image created and passed
to enclave as extended disk. The test case passes the details about this
new block disk image in the application configuration file. The test framework
will pick this image details in the environment variable and use it in test.
Test case changes link: lsds/ltp#49, lsds/ltp#59.

This PR also revert the PR:https://github.com/lsds/sgx-lkl/pull/741
This lsds/sgx-lkl#741, creates a filesystem image in master image.
because of this, each individual test case needs to be modified.

shaikshavali1 committed:
LTP: Pass ext4 block device to enclave

This test case causing oom killer to be invoked and causing
kernel panic. This is because the test case invokes test framework
function "tst_acquire_device" which creates a 256MB of .img file
and binds with a loop device. This is not supported because the
default size of LKL memory is set to 32M. This is kept low due to
the limited size of EPC (Enclave page cache) and to avoid page
cache being swapped out.
Also, the test case invokes "tst_mkfs" framework function which
inturn invokes mkfs utility command with the help of system() syscall.
It is recommended to use root file system, but, this test case
needs a read-only filesystem to test one of the sub-test cases.
we don't have a read-only file system mounted in sgx-lkl.

To support the test cases which need a block device a formatted
ext4 filesystem image created and passed
to enclave as extended disk. The test case passes the details about this
new block disk image in the application configuration file. The test framework
will pick this image details in the environment variable and use it in test.
Test case changes link: lsds/ltp#49, lsds/ltp#59.

This PR also revert the PR:https://github.com/lsds/sgx-lkl/pull/741
This lsds/sgx-lkl#741, creates a filesystem image in master image.
because of this, each individual test case needs to be modified.

shaikshavali1 committed:
Added enclave and host config files

shaikshavali1 committed:
Added enclave and host config files

shaikshavali1 committed:
Update enclave config file

shaikshavali1 committed:
Update enclave config file

hukoyu commented on tests/ltp/batch.mk:
```diff
@@ -22,7 +26,10 @@ ESCALATE_CMD=sudo
-$(ROOT_FS): $(ALPINE_TAR) ../buildenv.sh
+$(LTP_TEST_MNT_IMG):
+	dd if=/dev/zero of=$(LTP_TEST_MNT_IMG) count=$(LTP_TEST_MNT_IMG_SIZE) bs=1M
+
```
Why this doesn't have line
mkfs -t ext4 $(LTP_TEST_MNT_IMG)

shaikshavali1 replied:
done

hukoyu commented on tests/ltp/run_ltp_test.sh:
```diff
@@ -28,7 +28,10 @@ case "$test_mode" in
+SGXLKL_LTP_TSTAPP_CFG="--enclave-config=ltp_tstapp_enclave_config.json"
+SGXLKL_LTP_HOST_CFG="--host-config=ltp_host_config.json"
+
+SGX_LKL_RUN_CMD=( "$SGXLKL_STARTER" $SGXLKL_LTP_HOST_CFG $SGXLKL_LTP_TSTAPP_CFG $run_flag sgxlkl-miniroot-fs.img )
```
Please put all env variables in double quotes similar to "$SGXLKL_STARTER"

shaikshavali1 replied:
done

hukoyu reviewed (changes requested):
@shaikshavali1 you have missing command in batch.mk. Please address my comments and if it passes tests after these changes, in general it looks good.

Francis-Liu committed:
Check lkl_virtio_dev_setup() return value

shaikshavali1 committed:
Addressed the review comments

Added missing mkfs command in batch.mk
used "" for environment variables in run_ltp_test file.

shaikshavali1 committed:
Addressed the review comments

Added missing mkfs command in batch.mk
used "" for environment variables in run_ltp_test file.

prp committed:
Ensure that lthread lock is held when joining thread (#752)

* This fixes a race condition in which an exiting thread may be
  deallocated before it has finished exiting.

* Clean up thread trace output and code formatting

shaikshavali1 committed:
Fix CI pipeline issue

prefix config path with ../

shaikshavali1 committed:
Fix CI pipeline issue

prefix config path with ../

Commit:
Ignore fd when doing MAP_ANONYMOUS for mmap

It's considered best practice for portable applications to provide -1
as the fd to mmap when doing MAP_ANONYMOUS. However, for Linux compatibility,
we want to ignore the fd entirely.Ignore fd when doing MAP_ANONYMOUS for mmap

SeanTAllen commented:
@shaikshavali1 this will need to be rebased (somewhat carefully) against oe_port as a few changes have happened with these files. 

See the following PRs that are merged or likely will be before this one:

- https://github.com/lsds/sgx-lkl/pull/738
- https://github.com/lsds/sgx-lkl/pull/735
- https://github.com/lsds/sgx-lkl/pull/746

prp committed:
Dump stack traces from all lthreads on SIGTERM/USR1 (#751)

* Dump stack traces from all lthreads on SIGTERM and SIGUSR1
* Fix merge conflict for displaying mount tables
* Fix support for SGXLKL_TRACE_SIGNAL

hukoyu committed:
Enable 3 LTP tests: access02, set_robust_list01, preadv203 (#739)

hukoyu commented:
@shaikshavali1  as @SeanTAllen  said PR https://github.com/lsds/sgx-lkl/pull/746 is merged and you have to change your PR and make changes in batch.mk only according to new file/folder logic. Please rebase and reapply your changes.

shaikshavali1 committed:
Added new line

shaikshavali1 committed:
Merge branch 'mountissue_fix' and forked branch

shaikshavali1 commented:
> @shaikshavali1 as @SeanTAllen said PR #746 is merged and you have to change your PR and make changes in batch.mk only according to new file/folder logic. Please rebase and reapply your changes.

@hukoyu, I rebased changes.

SeanTAllen commented:
It looks like something is wrong with the rebase. There's a lot of unrelated changes in this PR now.

shaikshavali1 commented:
> It looks like something is wrong with the rebase. There's a lot of unrelated changes in this PR now.

I will raise the new PR. It looks it is not possible to fix it now

shaikshavali1 commented:
@hukoyu , This pull request is broken now. Raised a new PR https://github.com/lsds/sgx-lkl/pull/764. Please verify and close this.

hukoyu commented:
Dupe of https://github.com/lsds/sgx-lkl/pull/764
